### PR TITLE
Feature/pregel vertex accumulation algorithm 2 data access

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

Added methods (and tests): 
- writeVertex: Enable program machine execution for writing vertices
- readVertex, readEdge: Added "filter" to reduce copy edge and vertex documents

#### Related Information

- We still need more tests. 

### Testing & Verification

- [x] This PR adds tests that were used to verify all changes:
  - [x] Added **JS Tests**

To execute, use: 
```
 test = require("@arangodb/air/data-access"); test.test()
```

### Documentation

Needs to be written.

### CODE REVIEW

File: arangod/Pregel/Algos/AIR/GraphFormat.cpp [hunk 1]
```diff
@@ -35,37 +40,160 @@ GraphFormat::GraphFormat(application_features::ApplicationServer& server,
...
+    // copy only specified keys/key-paths to document
+    VPackBuilder tmpBuilder;
+    filterDocumentData(tmpBuilder, _dataAccess.readVertex->slice(), vertexDocument);
+    targetPtr.reset(_vertexAccumulatorDeclarations, _customDefinitions,
+                    _dataAccess, documentId, tmpBuilder.slice(), _vertexIdRange++);
+  } else {
+    // copy all
+    targetPtr.reset(_vertexAccumulatorDeclarations, _customDefinitions,
+                    _dataAccess, documentId, vertexDocument, _vertexIdRange++);
+  }
```
Comment (markuspf) [inline arangod/Pregel/Algos/AIR/GraphFormat.cpp:131]: note that we have full control over the type of `vertex_data`, so we could get away with not having yet another builder here.

File: arangod/Pregel/Algos/AIR/GraphFormat.cpp [hunk 2]
```diff
@@ -35,37 +40,160 @@ GraphFormat::GraphFormat(application_features::ApplicationServer& server,
...
-  targetPtr.reset(edgeDocument);
+  if (_dataAccess.readEdge->slice().isArray()) {
+    // copy only specified keys/key-paths to document
+    VPackBuilder tmpBuilder;
+    filterDocumentData(tmpBuilder, _dataAccess.readEdge->slice(), edgeDocument);
+    targetPtr.reset(tmpBuilder.slice());
+  } else {
+    // copy all
+    targetPtr.reset(edgeDocument);
+  }
```
Comment (markuspf) [inline arangod/Pregel/Algos/AIR/GraphFormat.cpp:143]: Same as above for `edge_type`

File: arangod/Pregel/Algos/AIR/GraphFormat.cpp [hunk 3]
```diff
@@ -35,37 +40,160 @@ GraphFormat::GraphFormat(application_features::ApplicationServer& server,
...
+                    auto&& [accumId] = res.value();
+
+                    if (auto iter = ptr->_vertexAccumulators.find(accumId);
+                        iter != std::end(ptr->_vertexAccumulators)) {
+                      return iter->second->getValueIntoBuilder(tmpBuilder);
+                    }
+                    return greenspun::EvalError("vertex accumulator `" + std::string{accumId} +
+                                                "` not found");
+                  });
+
```
Comment (markuspf) [inline arangod/Pregel/Algos/AIR/GraphFormat.cpp:169]: `vertexData` already has `air_accumRef`, can that be used here?

File: arangod/Pregel/Algos/AIR/GraphFormat.cpp [hunk 4]
```diff
@@ -35,37 +40,160 @@ GraphFormat::GraphFormat(application_features::ApplicationServer& server,
...
+    VPackBuilder tmpBuilder;
+    auto res = Evaluate(m, ptr->_dataAccess.writeVertex->slice(), tmpBuilder);
+    if (res.fail()) {
       LOG_DEVEL << "finalize program failed: " << res.error().toString();
       TRI_ASSERT(false);
```
Comment (markuspf) [inline arangod/Pregel/Algos/AIR/GraphFormat.cpp:174]: This should likely also `THROW_ARANGO_EXCEPTION`

File: js/client/modules/@arangodb/air/pregel-example-graphs.js [hunk 5]
```diff
@@ -206,7 +206,49 @@ function createCircle(graphName, numberOfVertices, numberOfShards) {
...
+          pathAsString = pathAsArray[pos];
+        } else if (pos < (pathAsArray.length - 1)) {
+          // middle iteration (object buildup)
+          pathAsString += '.' + pathAsArray[pos];
+        } else {
+          // last step (also insert value i here)
+          _.set(obj, pathAsString, )
+        }
+      }*/
+    }
```
Comment (markuspf) [inline js/client/modules/@arangodb/air/pregel-example-graphs.js:245]: What is this and why is it commented out?

File: arangod/Pregel/Algos/AIR/GraphFormat.cpp [hunk 6]
```diff
@@ -35,37 +40,160 @@ GraphFormat::GraphFormat(application_features::ApplicationServer& server,
                          std::string const& resultField,
                          AccumulatorsDeclaration const& globalAccumulatorDeclarations,
                          AccumulatorsDeclaration const& vertexAccumulatorDeclarations,
-                         CustomAccumulatorDefinitions customDefinitions)
```
[the remark this answers is not in the document]
Comment (markuspf) [reply arangod/Pregel/Algos/AIR/GraphFormat.cpp:38]: So I mean in `arangod/Pregel/Algos/AIR/AIR.cpp` line 80, just pass `*this` (i.e. a reference to the algorithm object, and make accessor functions for the declarations in there.

File: js/client/modules/@arangodb/air/data-access.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
+// //////////////////////////////////////////////////////////////////////////////
```
Comment (markuspf) [inline js/client/modules/@arangodb/air/data-access.js:1]: Yay tests!

File: arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp [hunk 8]
```diff
@@ -265,6 +266,113 @@ EvalResult Prim_Dict(Machine& ctx, VPackSlice const params, VPackBuilder& result
...
+
+    if (iter.isLast()) {
+      // if end of path reached, remove last visited member
+      if (currentPath.size() > 0) {
+        currentPath.pop_back();
+      }
+    }
+  }
+}
+
```
Comment (markuspf) [inline arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp:341]: Is this tested at the moment?

**Review by markuspf — APPROVED** [review]
Overall looks good to me, in particular having tests is good ;)

A  few comments about future  refactoring tasks that should ideally not get lost, but are not necessary for this to be merged right now.

File: js/client/modules/@arangodb/air/pregel-example-graphs.js (hunk 5 above)
Comment (hkernbach) [reply js/client/modules/@arangodb/air/pregel-example-graphs.js:245]: Ah, this code I wanted to use before I've found that there is the underscore set method :) 
Will remove it completely now.

File: arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp (hunk 8 above)
Comment (hkernbach) [reply arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp:341]: It is. That is being used within the dict-directory primitive. And that one is used in the JS test. But yes, we should add standalone primitive test as well!

**Pull request merged**